### PR TITLE
feat(manager/poetry): Support tag as git-dependency on github

### PR DESCRIPTION
**Pull request by yetanotherion**
## Changes

Parse git dependencies like
```
{git = "https://github.com/tiangolo/fastapi.git", tag="1.2.3"}
```
 or
```
{git = "git@github.com:tiangolo/fastapi.git", tag="1.2.3"}
```

and support them as github-tags datasources.

## Context
GitHub Packages currently lacks support for [PyPI repositories](https://github.com/github/roadmap/issues/94). As an alternative, some companies opt to use git-dependencies for hosting internal dependencies rather than installing third-party solutions such as [artifactory](https://jfrog.com/artifactory/) or Google’s [Artifact Registry](https://cloud.google.com/artifact-registry). This pull request adds support for specific use cases that rely on git tags, as discussed in issue https://github.com/renovatebot/renovate/issues/21848.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by viceice** [comment]
please correctly fill the PR template

File: lib/modules/manager/poetry/extract.ts
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:11]: you also need to update the `index.ts`. checke the supported datasources

File: lib/modules/manager/poetry/extract.ts
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:73]:
```suggestion
```

File: lib/modules/manager/poetry/extract.ts
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:84]:
```suggestion
        if (currentValue.tag) {
```

File: lib/modules/manager/poetry/extract.ts
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:85]:
```suggestion
          currentValue = currentValue.tag;
```

File: lib/modules/manager/poetry/extract.ts
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:221]: please try `parseGitUrl` instead

https://github.com/renovatebot/renovate/blob/df583b67995395e0be3701cbb42c8d89f11cbe7f/lib/util/git/url.ts#L7-L9

File: lib/modules/manager/poetry/extract.ts
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:228]: no nested functions

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/poetry/extract.spec.ts
Comment (rarkins) [inline lib/modules/manager/poetry/extract.spec.ts:183]: Using `codeblock` syntax is the preferred way now - easier to read and maintain

File: lib/modules/manager/poetry/extract.ts
Comment (rarkins) [inline lib/modules/manager/poetry/extract.ts:69]: Was there a need to refactor/move this logic?

Comment (yetanotherion) [reply lib/modules/manager/poetry/extract.ts:69]: Unfortunately yes, we change the packageName to for ex `tiangolo/fastapi.git` when depName and the key in the poetryLockfile is `fastapi`.

File: lib/modules/manager/poetry/extract.spec.ts
Comment (yetanotherion) [reply lib/modules/manager/poetry/extract.spec.ts]: Did I understand the suggestion correctly? (https://github.com/renovatebot/renovate/pull/21949/commits/81364455785463b63664d62c9c925a83b33e2205)
If so, would you like me to refactor remaining lines in that test file too? (Don't hesitate, I'd be glad to do so)

Comment (rarkins) [reply lib/modules/manager/poetry/extract.spec.ts]: It would be great if you could, thanks. It makes it a little harder for us to verify that no tests have "changed" (logically) but would be worth it. FYI @viceice

Comment (yetanotherion) [reply lib/modules/manager/poetry/extract.spec.ts]: I did it at https://github.com/renovatebot/renovate/pull/21949/commits/508e83cecaaf99b654b2d40fd4c3a1a9d12d2493
and will revert the commit if it makes it harder to review.

File: lib/modules/manager/poetry/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/poetry/extract.spec.ts:145]:
```suggestion
      const content = codeBlock`
```
then you can properly align code indented and the helper will remove leading whitespace

Comment (yetanotherion) [reply lib/modules/manager/poetry/extract.spec.ts]: Thanks viceice! I initially ran 
```
git grep codeblock
```
or asked google about it and did not find anything.
```
git grep codeBlock
```
gave me the examples so that I could be a very precise model of what is the current way of writing those.
(https://github.com/renovatebot/renovate/pull/21949/commits/353164341c5ffb09c4e17b7f838bc51559ec0029)

**Review by viceice — APPROVED** [review]

**Comment by yetanotherion** [comment]
Thanks viceice! I'll wait for rarkins's approval before merging.

**Review by rarkins — APPROVED** [review]

**Comment by yetanotherion** [comment]
Thanks a lot rarkins!

**Comment by yetanotherion** [comment]
rarkins, viceci, just wanted to congratulate you for the process that you guys put in place in renovate.

Everything is so self explanatory that I don't even have to ask when the PR will be released:
- I have an automated comment that [tells me so](https://github.com/renovatebot/renovate/pull/21949#issuecomment-1533423276)
- With the [info](https://github.com/renovatebot/github-action/releases) in the releases of the related github-action, I can conclude that this feature will be available in the next days in the github-action too.

So far, contributing to renovate has been **the best experience** I ever had working in an open source project 👏 👏 👏, thank you!

**Pull request merged**
